Increase test coverage in pkg/sypgp

mem opened the pull request:
- Break up PathsCheck so that the individual pieces can be tested
- Add ensureDirPrivate and ensureFilePrivate which are testable using
  temporary directories
- Add tests for ensureDirPrivate and ensureFilePrivate
- Add PrintEntity test

Signed-off-by: Marcelo E. Magallon <marcelo@sylabs.io>

cclerget commented on pkg/sypgp/sypgp.go:
```diff
@@ -98,61 +105,67 @@ func PublicPath() string {
+func ensureDirPrivate(dn string) error {
+	mode := os.FileMode(0700)
+
+	if err := os.MkdirAll(dn, mode); err != nil {
```
There is almost no chance that a fancy umask alter those permissions here but just in case I would reset umask before creating directories/files by precaution

mem replied:
yikes ... I tried to _only_ move code around, with minimal fixes, but given how what this code is doing is already not so nice, I'll do as suggested, thanks.

cclerget commented on pkg/sypgp/sypgp.go:
```diff
@@ -98,61 +105,67 @@ func PublicPath() string {
+func ensureFilePrivate(fn string) error {
+	mode := os.FileMode(0600)
+
+	fs, err := os.OpenFile(fn, os.O_RDWR|os.O_CREATE, mode)
```
same here

mem replied:
this is resolved, dismissing in order to be able to move.

Commit:
Add test for CompareKeyEntity

Signed-off-by: Marcelo E. Magallon <marcelo@sylabs.io>

Commit:
Add test for findKeyByFingerprint

Signed-off-by: Marcelo E. Magallon <marcelo@sylabs.io>

Commit:
Add createOrAppendPrivateFile to reduce code duplication

Several code paths need to either create a file or append to it if it
already exists, and they need to create it with a specific mode.

Move this to a function in order to avoid code duplication.

Signed-off-by: Marcelo E. Magallon <marcelo@sylabs.io>

Commit:
Move storing keys to common functions

Signed-off-by: Marcelo E. Magallon <marcelo@sylabs.io>

Commit:
Add removeKey to handle removal of keys

Given a list of keys (entities) and a fingerprint, removeKey will remove
the matching key from the provided list.

Signed-off-by: Marcelo E. Magallon <marcelo@sylabs.io>

Commit:
Rework RemovePubKey to use the new functions

Even if testing RemovePubKey itself is hard because it's operating on
the filesystem, it should be possible to test the individual operations.

Signed-off-by: Marcelo E. Magallon <marcelo@sylabs.io>

Commit:
Move common code out of SelectPubKey / SelectPrivKey

Signed-off-by: Marcelo E. Magallon <marcelo@sylabs.io>

Commit:
Rework RecryptKey to take a passphrase

In order to make RecryptKey non-interactive, move the logic that asks
for a passphrase outside of it.

Produce an error if the key that needs reencrypting is not encrypted.

Signed-off-by: Marcelo E. Magallon <marcelo@sylabs.io>

Commit:
Move defer file.Close() closer to where it's opened

Signed-off-by: Marcelo E. Magallon <marcelo@sylabs.io>

WestleyK commented on pkg/sypgp/sypgp.go:
```diff
@@ -376,144 +479,136 @@ func GetPassphrase(message string, retries int) (string, error) {
+	ans, err := askYNQuestion("y", "Would you like to push it to the keystore? [Y,n] ")
+	switch {
+	case err != nil:
+		fmt.Printf("Not pushing newly created key to keystore: %s\n", err)
```
Should this print to stderr?

```suggestion
		fmt.Fprintf(os.Stderr, "Not pushing newly created key to keystore: %s\n", err)
```

WestleyK commented on pkg/sypgp/sypgp.go:
```diff
@@ -376,144 +479,136 @@ func GetPassphrase(message string, retries int) (string, error) {
 // DecryptKey decrypts a private key provided a pass phrase
 func DecryptKey(k *openpgp.Entity, message string) error {
+	if !k.PrivateKey.Encrypted {
+		return errNotEncrypted
```
Why return a error if not encrypted?

mem replied:
my expectation is that if you are calling the DecryptKey function it's because you think the key is encrypted. If the key is not encrypted, why are you calling DecryptKey?

WestleyK replied:
What if you remove the `if !k.PrivateKey.Encrypted {...}`, then when we want to decrypt a key later we can do:

```go
if k.PrivateKey.Encrypted {
    err := sypgp.DecryptKey(entity, pass)
    if err != nil {
        sylog.Fatalf("Unable to decrypt key: %s", err)
    }
}
```

Of course you can check `k.PrivateKey.Encrypted` anyway, but then `k.PrivateKey.Encrypted` will be called twice.

WestleyK commented on pkg/sypgp/sypgp_test.go:
```diff
@@ -6,16 +6,20 @@
 package sypgp
 
```
```
// Copyright (c) 2018-2019, Sylabs Inc. All rights reserved.
```

Update copyright year in this file.

WestleyK reviewed (commented):
Couple suggestions/questions, otherwise looks really good.

bauerm97 reviewed (approved):
I gave it a 10 minute look through, structure of the code looks good to me. As long as CI is passing, I don't see any glaring reasons why we shouldn't merge

WestleyK commented on pkg/sypgp/sypgp.go:
```diff
@@ -376,144 +479,136 @@ func GetPassphrase(message string, retries int) (string, error) {
+	printEntities(os.Stdout, el)
 
-	index, err := AskQuestion("Enter # of signing key to use : ")
+	n, err := askNumberInRange(0, len(el)-1, "Enter # of signing key to use : ")
```
```suggestion
	n, err := askNumberInRange(0, len(el)-1, "Enter # of private key to use : ")
```

Because when you export a key.

mem replied:
Done, thanks.

WestleyK commented on pkg/sypgp/sypgp.go:
```diff
@@ -376,144 +479,136 @@ func GetPassphrase(message string, retries int) (string, error) {
 
 // SelectPrivKey prints a secret key list to user and returns the choice
 func SelectPrivKey(el openpgp.EntityList) (*openpgp.Entity, error) {
-	PrintPrivKeyring()
```
Please add:

```go
if el == nil {
    return nil, fmt.Errorf("no private keys in local keyring")
}
```

To avoid a panic. _and same for `SelectPubKey`._

mem replied:
I see this with an empty keyring:

```
$ > /home/mem/.singularity/sypgp/pgp-public
$ builddir/singularity key list
Public key listing (/home/mem/.singularity/sypgp/pgp-public):

$ > /home/mem/.singularity/sypgp/pgp-secret
$ builddir/singularity sign alpine_3.9.sif
Signing image: alpine_3.9.sif
ERROR:   signing container failed: no private keys in keyring. use 'key newpair' to generate a key, or 'key import' to import a private key from gpg
```

WestleyK replied:
First, I made a PR for this change https://github.com/sylabs/singularity/pull/3857 (bug fix PR). Second, it will only panic if you try to export a key.

WestleyK replied:
Probably best to just ignore this and it will be fixed in the bug-fix PR. Better to keep bug-fixes separate with testing PRs.

WestleyK reviewed (changes requested):
Left some suggestions, and this comment: https://github.com/sylabs/singularity/pull/3290#discussion_r278761615

WestleyK reviewed (commented):
Left a comment: https://github.com/sylabs/singularity/pull/3290#discussion_r300819863, just ignore my suggestion (it will be fixed in a separate bug-fix PR). Otherwise looks good.

Commit:
Merge branch 'master' into sypgp_increased_coverage
